[RAD-2751 pt2] Use micro nor cal region for all unit + functional tests

mark-idleman opened the pull request:
Part 2 of [RAD-2751](https://replicahq.atlassian.net/browse/RAD-2751). Moves us fully over to using micro_nor_cal as a test region, both for unit tests + our current functional test. This requires integrating the use of the `setup_tests.sh` shell script that was added in #77 . In particular, this PR:
- Integrates `setup_tests.sh` directly into CI, ensuring that the container we run `mvn test` in for unit tests + the container we run functional tests against contain the proper micro_nor_cal test files
- Adds a script `check_test_setup.sh` to ensure that the setup script has been run before tests get kicked off. Only runs when running a mvn lifecycle command that includes the test phase, and doesn't run if `DskipTests` is set (ie, in `Dockerfile.build`, where we don't expect test fixtures to be setup properly yet). The goal of this script is to remind you locally to run the setup script if you're trying to run unit tests. Unfortunately, forcing a run of `setup_tests` itself pre-maven-test doesn't work with our CI test container, so that approach isn't possible
- Migrates unit + functional test logic to test new micro_nor_cal region. A few new unit tests are added - different tests to check inter- and intra-feed transfers, and a test for a "bad" PT request, now possible because of Michael's recent fix (these would hang ♾️ before)
- Sets up a junit extension containing the setup + teardown code that all of the unit tests currently share; this is the only way to force the `import` + `gtfs_links` steps to be run only once before _all_ tests get kicked off, instead of before each individual test/test file (the default junit behavior if you just use annotations like `@beforeall`). This saves us a bunch of time, because graph building + link mapping takes ~5 mins

cc @JacobHayes @epotex @danielhfrank

mark-idleman committed:
first pass at getting mini nor cal fixture data downloaded in circle

mark-idleman committed:
test-data not test_data

mark-idleman committed:
try ./

mark-idleman committed:
add more .s

mark-idleman committed:
try pulling gtfs file list logic out of sed command

mark-idleman committed:
oops

mark-idleman committed:
try this

mark-idleman committed:
try double quotes for sed

mark-idleman committed:
escape / properly in gtfs file list

mark-idleman committed:
tiny tweak

mark-idleman committed:
remove cat of config file now that mini nor cal download is working

mark-idleman committed:
remove test config

mark-idleman committed:
switch test OD + departure time to mini_nor_cal

mark-idleman committed:
Revert "remove test config"

This reverts commit b84817a3f98aa42e2e65c2b946a7172c77c7bbb1.

mark-idleman committed:
more cleanup/mini_kc removal; add setup script, redo circle steps

mark-idleman committed:
remove pipefail option

mark-idleman committed:
fix setup script

mark-idleman committed:
try this

mark-idleman committed:
try this #2

mark-idleman committed:
attempt to fix docker cp

mark-idleman committed:
try again

mark-idleman committed:
fix again

mark-idleman committed:
try this

mark-idleman committed:
add -t

mark-idleman committed:
try #54321

mark-idleman committed:
mkdir in running container

mark-idleman committed:
try with .

mark-idleman committed:
correct path by adding /web

mark-idleman committed:
fix sed command

mark-idleman committed:
add . to log

mark-idleman committed:
OOPS

mark-idleman committed:
try fixing filepaths in test container

mark-idleman committed:
try try try try

mark-idleman committed:
try again

mark-idleman committed:
fix

mark-idleman committed:
fix after testing locally

mark-idleman committed:
update script; bump heap

mark-idleman committed:
upgrade setup script to work both locally + in CI, plus check for existing files

mark-idleman committed:
try using subset of GTFS

mark-idleman committed:
oops

mark-idleman committed:
fix script

mark-idleman committed:
update to cutout mini nor cal OSM too

mark-idleman committed:
fix linux osmtools install

mark-idleman committed:
try -y

mark-idleman committed:
another mini -> micro

mark-idleman committed:
fix tests to work with micro nor cal; setup script run during mvn test; run setup/teardown once only

mark-idleman committed:
export path vars

mark-idleman committed:
try again

mark-idleman committed:
TEST echo

mark-idleman committed:
update circle setup to work with automated setup script

mark-idleman committed:
skip mvn exec if DskipTests is set

mark-idleman committed:
run tests passing osm/gtfs path env vars

mark-idleman committed:
revert back to not forcing setup_tests to run on mvn test

mark-idleman committed:
merge master

mark-idleman committed:
add script for checking if test setup has been run

mark-idleman committed:
try adding setup script to gh actions CI

mark-idleman committed:
make setup check executable

mark-idleman committed:
fix check script

mark-idleman committed:
try exporting secrets as env vars before running setup

mark-idleman committed:
try sudo to make gh actions happy

mark-idleman committed:
tweaks to (hopefully) fix both CIs

mark-idleman committed:
add echos for testing

mark-idleman committed:
try removing -t in  gh actions config

mark-idleman committed:
try only using -t for run

mark-idleman committed:
cleanup echos; try fixing functional test

mark-idleman commented on .github/workflows/build_test_push.yaml:
```diff
@@ -62,8 +62,21 @@ jobs:
+      - name: run unit tests
+        run: |
+          docker run -t -d --name testrunner us.gcr.io/model-159019/gh:${{ github.sha }}-dev /bin/bash
+          docker exec -i testrunner /bin/bash -c "mkdir -p ./web/test-data/gtfs"
```
note here: for some reason, running `exec` with `-it` was complaining at me here; removing the `-t` made things work. Not an issue in circle, so didn't change anything there 🤷

mark-idleman commented on setup_tests.sh:
```diff
@@ -23,7 +23,7 @@ if [[ ! -f "./web/test-data/micro_nor_cal.osm.pbf" ]]; then
       brew install osmfilter
     else
-      apt install -y osmctools
+      sudo apt install -y osmctools
```
note: gh actions made me add this, basically. This command worked fine in circle but I think permissions are different in the actions setup

mark-idleman committed:
copy test gh config into server image too, for func test

mark-idleman committed:
remove unnecessary commit

kaelgreco commented on .circleci/config.yml:
```diff
@@ -27,9 +27,21 @@ jobs:
           name: Run Tests
-          command: docker run us.gcr.io/model-159019/gh:$(cat /tmp/tag)_build /bin/bash -c "mvn test"
+          command: |
+            docker run -t -d --name testrunner us.gcr.io/model-159019/gh:$(cat /tmp/tag)_build /bin/bash
```
ahh i see what's happening now. some of the docker commands are a little funky. This is mostly to avoid keeping the test data in the repo? (How big is it?) I'm happy if this works, but out of curiosity is there any way to encapsulate this in a test dockerfile that used the base image?

mark-idleman replied:
See my other comment for more reasoning to why I took this approach.

Hmmmm, now that you mention it, though....micro_nor_cal could probably fit in GH directly 😳 . OSM is only ~40mb...so, mini_nor_cal wasn't possible, but micro probably is. Perhaps that changes the approach here, I guess we could just generate it manually with the script and plop it in the repo....

re `is there any way to encapsulate this in a test dockerfile that used the base image?`, what do you mean exactly? An image that contains the test fixture data + has access to `mvn` to run tests against? That's essentially what we're building with the commands I have in this section - a few changes and we could push the result, then run `mvn test` against it (instead, we're just "building" it live in a container and then running the test command)

kaelgreco replied:
I think it's fine to move forward with this. we can reorganize in the future if necessary. I was trying to think of a better way to encapsulate the ci commands in a new test dockerfile to streamline the circleconfig.

mark-idleman commented on test_gh_config.yaml:
```diff
@@ -2,8 +2,8 @@
 graphhopper:
-  datareader.file: test-data/kansas-city-extract-mini.osm.pbf
-  gtfs.file: test-data/mini_kc_gtfs.tar
+  datareader.file: TEST_OSM
```
errr I'm not sure - I think the only reason we need curly braces in the python code is because python's `format_map()` requires them. Here, we're just using `sed`, which is really direct regex replacement. In my mind, it's almost more confusing to find + replace `{TEST_OSM}` vs just `TEST_OSM`?

kaelgreco replied:
I just wanted it for readability (to connote that `TEST_OSM` would be overwritten). Looking around maybe a double `{{}}` would be best? (single curly braces designate associative arrays).

mark-idleman commented on pom.xml:
```diff
@@ -45,11 +45,31 @@
             </plugin>
+            <plugin>
+                <groupId>org.codehaus.mojo</groupId>
+                <artifactId>exec-maven-plugin</artifactId>
```
almost! The plugin executes any time the `test` phase in maven runs (for our purposes, that's any time you run `mvn package` or `mvn test`). It does _not_ run if you provide the `-DskipTests` flag (as we do in `Dockerfile.build` - where we don't have the test data ready to run tests - and as is pretty common locally when running `package`).

It's basically a gate to ensure that any time maven would be running tests, we double check that the test files are ready first, and end things if not. 

My initial approach was to not run a check at all, but just to run the `setup_tests` script itself in these same situations, which would have alleviated all the crazy stuff I'm doing in the CI [here](https://github.com/replicahq/graphhopper/pull/78#discussion_r680282376). This seems so close to working and much more slick - we just run `mvn test` against the container and any necessary setup runs automatically beforehand! But the problem is that the container at that point only has a mvn base image to work with at that point, and can't run all of the commands in the setup script ([ex](https://app.circleci.com/pipelines/github/replicahq/graphhopper/788/workflows/e8128421-6703-47d0-aac9-04763e9b387c/jobs/832)). Perhaps we can install gsutil in that layer of the image to make this work, but it just feels messy

mark-idleman commented on functional_test.sh:
```diff
@@ -25,18 +25,16 @@ docker run \
-     -Ddw.graphhopper.datareader.file=test-data/kansas-city-extract-mini.osm.pbf \
-     -Ddw.graphhopper.gtfs.file=test-data/mini_kc_gtfs.tar \
-     -classpath web/target/graphhopper-web-1.0-SNAPSHOT.jar -server com.graphhopper.http.GraphHopperApplication import default_gh_config.yaml
+     java -Xmx4g -Xms1g -XX:+UseG1GC -XX:MetaspaceSize=100M \
```
Yup, I think so. I can confirm again though, probably good to be sure I wasn't setting this for full mini_nor_cal

kaelgreco reviewed (approved):
very excited for this!

mark-idleman committed:
{{ }}

mark-idleman committed:
FOR TESTING ONLY try 2g heap again for funct test

mark-idleman committed:
merge original-direction

mark-idleman commented:
Addressed PR comments and got a lower ram setting (2g instead of 4) to work with the functional tests, so we can go with that, woohoo!

mark-idleman committed:
TEMPORARY force run func test to check 2g ram works

mark-idleman committed:
Revert "TEMPORARY force run func test to check 2g ram works"

This reverts commit d0cc1d6af4a3fa5217655a7736fd1886ad807c3a.

mark-idleman commented:
Addressed PR comments and [confirmed](https://github.com/replicahq/graphhopper/pull/78/commits/d0cc1d6af4a3fa5217655a7736fd1886ad807c3a) that less ram (2g) will work for functional tests!
